### ISSUE TITLE
'change' takes forever depending on Opaque status

### ISSUE THREAD

**Issue by andrew-appel** [coq/coq#7316]
#### Operating system: any

In the quoted example below, the `change` command takes forever, or is instantaneous, depending on interactions between the Transparent/Opaque status of two different definitions.  In general, we expect that `change` will disregard Opaque.  (This is in contrast to issue #6998, where we want `simple apply` to avoid unfolding Opaque definitions.)

I am not sure that this is a bug; it _feels_ like a bug.  It is not critical, as I have already found a workaround in the particular case.  However, this kind of behavior causes Ltac programs to be unpredictable, and is therefore undesirable.    Thanks to @Aurele-Barriere for helping to discover and isolate this issue.

```
Require Import ZArith.
Open Scope Z.

Fixpoint P_mod_two_p (p: positive) (n: nat) {struct n} : Z :=
  match n with
  | O => 0
  | S m =>
      match p with
      | xH => 1
      | xO q => Z.double (P_mod_two_p q m)
      | xI q => Z.succ_double (P_mod_two_p q m)
      end
  end.

Definition Z_mod_modulus (x: Z) : Z :=
  match x with
  | Z0 => 0
  | Zpos p => P_mod_two_p p 32
  | Zneg p => 0 (* bogus value *)
  end.

Record int: Type := mkint { intval: Z }.
Definition repr (x: Z) : int := mkint (Z_mod_modulus x).

Definition f := intval.

Lemma test: forall x, 0 <= intval (repr (Zpos ( x))).
Proof.
  intros.
(* f transparent, Z_mod_modulus transparent *)
Fail Timeout 10 change (intval (repr (Zpos x))) with (f (repr (Zpos x))). (* times out *)
(* f opaque, Z_mod_modulus transparent *)
Opaque f.
Fail Timeout 10 change (intval (repr (Zpos x))) with (f (repr (Zpos x))). (* times out *)
(* f opaque, Z_mod_modulus opaque *)
Opaque Z_mod_modulus.
Fail Timeout 10 change (intval (repr (Zpos x))) with (f (repr (Zpos x))). (* times out *)
Transparent f.
(* f transparent, Z_mod_modulus opaque *)
Time  change (intval (repr (Zpos x))) with (f (repr (Zpos x))).  (* 0.0 seconds *)
```

**Comment by maximedenes** [comment]
> In general, we expect that change will disregard Opaque.

Hi Andrew, where does this expectation come from? Citing the manual: " The [Opaque] command tells not to unfold the constants in the qualid sequence in tactics using δ-conversion". Why wouldn't it apply to `change`?

Or are you suggesting a new semantics for `Opaque`? If so, what would it be?

**Comment by andrew-appel** [comment]
I have _always_ used `change`  to fold/unfold Opaque definitions.  This has been the way Coq works for at least 10 years, probably more.  It would be a very major incompatibility if `change` were to block on Opaque.

**Comment by maximedenes** [comment]
I see. Unfortunately, looking at the code, the current situation seems a bit random, with tactics starting to bypass `Opaque` at arbitrary places. Do you have a mental model of what tactics you expect to bypass `Opaque`? We could try to extract a new spec from it, and fix the remaining bugs.

**Comment by andrew-appel** [comment]
That's a good question.  My mental model of Ltac has always been, "one can never predict what works, one must do experiments."   One learns Ltac like one learns French (or conversely, English).  Obviously, that's not a good situation.  
   More specifically, we could experiment with,  some tactics respect Opaque (simple apply), some tactics bypass Opaque (change).  As discussed in issue #6998, it would probably be a good idea to put `apply` in the "respect Opaque" category, but that will cause some incompatibilities, so we should step carefully here.   I suggest that "reflexivity" should probably be in the "bypass Opaque" category, but I am not sure (what does the code do now?).   Regarding the `unify` tactic, I have no idea which is best--what does it do now?.

  Beyond the question of "which tactics respect/bypass Opaque," there is another problem here that is even worse:  tactics that infinite-loop (or exponentially loop) when confronted with opaque definitions (it doesn't matter whether they are half-opaque like Opaque or fully opaque like Qed).

**Comment by maximedenes** [comment]
A related question: in some of these reports (including the example above), you seem to actually need first-order unification/conversion tactics. That is, you seem to be using `Opaque` to restrict `apply` or `change` to first-order strategies. If we provided such tactics, would it help on the real examples?

**Comment by andrew-appel** [comment]
I use `Opaque` to prevent `simpl` from unfolding a definition.  I use `change` when I want to explicitly fold or unfold the definition.  I propose to preserve this power of `change`.

Then:  `change foo with bar` should work iff `replace foo with bar; [ | reflexivity]` works.  I believe that is mostly true now.  If we make `change` uniformly bypass Opaque, and make `reflexivity` uniformly bypass Opaque, we get consistent behavior between `change` and `reflexivity`, and we get almost 100% compatibility with Coq's current behavior.

Then, if you consider it desirable that `apply eq_refl` should mean exactly the same as `reflexivity`, we could allow `apply` to bypass Opaque as it does now.  I think that would be OK, as long as I have some Opaque-respecting way of applying lemmas:  that is, `simple apply` should respect Opaque.

Regarding, "is this just a behavior of the calculus", I expect the answer is "yes".

Finally: new tactics that do only first-order unification?  I am not sure what those would look like.

**Comment by maximedenes** [comment]
> I use Opaque to prevent simpl from unfolding a definition.

Isn't `Arguments foo : simpl never` better suited for this use case? Note that despite its name, it doesn't prevent all unfolding of `foo` by `simpl`, but we are also considering to add a flag that does it. So I would be interested in knowing if the current `simpl never` is enough for you or not.

**Comment by andrew-appel** [comment]
I absolutely need to prevent `simple apply` from unfolding Opaque definitions.  Are you proposing that `simple apply` will not unfold `foo` if I say `Arguments foo : simpl never`  ?

**Comment by maximedenes** [comment]
> I absolutely need to prevent simple apply from unfolding Opaque definitions. Are you proposing that simple apply will not unfold foo if I say Arguments foo : simpl never ?

No, sorry if I wasn't clear, I was suggesting to use `simpl never` when you need to control only `simpl` for a specific constant. If you need to also control `simple apply`, then yes, you need to use `Opaque`.

> I use change when I want to explicitly fold or unfold the definition.

Could we make `unfold`/`fold` bypass `Opaque` for this purpose, instead? Or would that require too many manual unfolding steps compared to `change`?

**Comment by ppedrot** [comment]
Why not simply provide a variant of `change` that disregards opaque? The kernel is always allowed to do this anyways, and `change` is one of the few tactics where it makes sense to have both (i.e. a kernel-based one disregarding opacity, and a unification-based one respecting it).

**Comment by JasonGross** [comment]
I rely on the fact that kernel conversion respects `Strategy` commands, including `Opaque`.  Please don't remove this knob for fine-tuning the conversion algorithm.

**Comment by ppedrot** [comment]
@JasonGross The kernel does not respect `Opaque`, at least in a way that unification respects it. It's just a hint that postpones unfolding as much as possible, while for unification it's really similar to `Qed` (modulo bugs).

**Comment by JasonGross** [comment]
I believe this bug, though, is specifically about the fact that conversion in the `change` tactic treats `Opaque` as a hint to delay unfolding.

I continue to believe that we should have the following knobs available:
- `Qed`-level-opaque, kernel never unfolds, doing this can break subject reduction, maybe mark it as pseudo-unsafe (you can't break consistency with it, but you can break subject reduction)
- current-`Opaque` : respected by `unfold`, `simpl`, `cbv`, etc.  In general respected by all reduction tactics, and treated as a hint to unfold as late as possible in conversion.  For tactics that fall somewhere between reduction and conversion (such as `setoid_rewrite`'s unfolding, maybe unification in general), have a flag that determines whether or not `Opaque` is taken as a hint or a block.
- all the numeric strategy levels
- `expand` (unfold first in conversion)
- some unfold-me-anytime-you-see-me-anywhere, where conversion and reduction and unification will always unfold this constant anytime they encounter it

(I'd be inclined to close this bug as invalid, btw, or perhaps turn it into a doc bug where this behavior should be better documented.)
